src/libcore/alloc/mod.rs

```diff
@@ -0,0 +1,337 @@
+    /// appropriate from `ptr`, and the remaining bytes, from `layout.size()` to the returned
+    /// actual size, are initialized according to [`init`].
+    ///
+    /// [`placement`]: ReallocPlacement
+    /// [`init`]: AllocInit
+    ///
+    /// # Safety
+    ///
+    /// * `ptr` must be [*currently allocated*] via this allocator,
+    ///
+    /// * `layout` must [*fit*] the `ptr`. (The `new_size` argument need not fit it.)
+    ///
+    // We can't require that `new_size` is strictly greater than `layout.size()` because of ZSTs.
+    // An alternative would be
+    // * `new_size must be strictly greater than `layout.size()` or both are zero
+    /// * `new_size` must be greater than or equal to `layout.size()`
+    ///
+    /// * `new_size`, when rounded up to the nearest multiple of `layout.align()`, must not overflow
+    ///   (i.e., the rounded value must be less than `usize::MAX`).
+    ///
+    /// [*currently allocated*]: #currently-allocated-memory
+    /// [*fit*]: #memory-fitting
+    ///
+    /// # Errors
+    ///
+    /// Returns `Err` if the new layout does not meet the allocator's size and alignment
+    /// constraints of the allocator, or if growing otherwise fails.
+    ///
+    /// Implementations are encouraged to return `Err` on memory exhaustion rather than panicking or
+    /// aborting, but this is not a strict requirement. (Specifically: it is *legal* to implement
+    /// this trait atop an underlying native allocation library that aborts on memory exhaustion.)
+    ///
+    /// Clients wishing to abort computation in response to an allocation error are encouraged to
+    /// call the [`handle_alloc_error`] function, rather than directly invoking `panic!` or similar.
+    ///
+    /// [`handle_alloc_error`]: ../../alloc/alloc/fn.handle_alloc_error.html
+    unsafe fn grow(
+        &mut self,
+        ptr: NonNull<u8>,
+        layout: Layout,
+        new_size: usize,
+        placement: ReallocPlacement,
+        init: AllocInit,
+    ) -> Result<(NonNull<u8>, usize), AllocErr> {
+        let old_size = layout.size();
+        debug_assert!(
+            new_size >= old_size,
+            "`new_size` must be greater than or equal to `layout.size()`"
+        );
+
+        if new_size == old_size {
+            return Ok((ptr, new_size));
+        }
+
+        intrinsics::assume(new_size > old_size);
+        match placement {
+            Unspecified => {
+                let (new_ptr, alloc_size) =
+                    self.alloc(Layout::from_size_align_unchecked(new_size, layout.align()), init)?;
+                debug_assert_eq!(cmp::min(old_size, new_size), old_size);
+                ptr::copy_nonoverlapping(ptr.as_ptr(), new_ptr.as_ptr(), old_size);
+                self.dealloc(ptr, layout);
+                Ok((new_ptr, alloc_size))
+            }
+            InPlace => Err(AllocErr),
+        }
+    }
+
+    /// Returns a pointer and the actual size of the allocated block. The pointer is suitable for
```

Comment:
The first sentence of the doc comment should explain what the method actually *does*. For example: "Attempts to shrink the allocation referenced by `ptr` to fit `new_size`."